### DISCUSSION TITLE
How to reduce build size in PyInstaller ?

### DISCUSSION THREAD

**Discussion by gaamaaresosa** [pyinstaller/pyinstaller#7587] [PyInstaller] [answered]
My appologize for this common question might have asked several times.
But nothing was clear to me yet.

My Python app using few of my custom python models.
Apart from that, I use lot of system installed imports also.
When I build my project with PI I see a huge size output.
I know PI is bundling all the system installed libs as the self contained copy for independent running.

I need PyInstaller should bundle only my custom libs and other libs, it has to use from the system installed path.
When I already have all those system libs on all my systems, why to bundle that again ?

1) I can exclude the system installed libs and make the build size smaller right ?

2) Why I see numpy libs are always added where on my project I didn't use numpy.
Since I am not an expert in linux, is numpy is mandatory for all python projects ?
If no can I exclude that too in PI ?

Btw, I don't use  Virtual Environments (venv) as I may app running on a Raspberry Pi on headless mode.

Thanks in advance !

**Comment by bwoodsend** [comment]
Virtual environments with only the packages that you need are the best way to go. That way, if PyInstaller sees packages with optional dependencies, it knows to exclude the ones you don't have installed.

> I can exclude the system installed libs and make the build size smaller right ?

Only if you know what you're doing. If the person you give your software to doesn't have that system library installed, your application will fail. Likewise if they have a significantly different version of that library.

> Why I see numpy libs are always added where on my project I didn't use numpy.
Since I am not an expert in linux, is numpy is mandatory for all python projects ?
If no can I exclude that too in PI ?

If you look inside `build/yourproject/xref-yourproject.html`, then search for numpy then it'll show you which module PyInstaller thought depended on numpy. If you're not using numpy then by all means use `--exclude=numpy` to get rid of it.

> Btw, I don't use Virtual Environments (venv) as I may app running on a Raspberry Pi on headless mode.

That's no reason not to use venv.

**Reply by gaamaaresosa** [reply]
@bwoodsend

Your answers are really amazing and any intermediates can understand easily.
Thank you so much.

>>Virtual environments with only the packages that you need are the best way to go

Great understood, now let me convert my project to a venv.

>> If the person you give your software to doesn't have that system library installed, your application will fail. 

My case is a custom made SD card with all preinstalled for Raspberry Pi 4
So customer don't have any access to the system.
Only web interface.

>>If you look inside build/yourproject/xref-yourproject.htm

Great idea, I will try that and use --exclude=numpy

NOTE:
I wonder why PyInstaller is not scaning all python files (*.py) get an import list and include only those packs on the PyIntaller package to reduce the size ?
In my new Cythonized PyInstaller (**CyInstaller**) project I scan and pass that list to PyInstaller.
It works fine.
But after using PyInstaller the output size become huge.
That's what I am working now.

I agree this is a safe way of packing to run the executable on any system without the required packages installed.
But there should be a **single flag option** to make light weight PyIntaller packages too right ?

**Reply by bwoodsend** [reply]
>  I wonder why PyInstaller is not scaning all python files (*.py) get an import list and include only those packs on the PyIntaller package to reduce the size ?

That is exactly what PyInstaller **does** do. Unfortunately, packages often contain imports in parts of the code you don't run or bits of their test suite (which imports test dependencies) which PyInstaller has no way of knowing that you're not going to need them.

>  But there should be a single flag option to make light weight PyIntaller packages too right ?

As long as you understand that if you put them on another computer without the dependencies then they're *light weight broken packages* then there's an [exclude_system_libraries](https://pyinstaller.org/en/v5.10.1/spec-files.html#posix-specific-options) spec file option. If you're shipping the whole OS on an SD card then it should be safe for you to use it.

**Reply by gaamaaresosa** [reply]
>>PyInstaller has no way of knowing that you're not going to need them.

Agreed.

>>If you're shipping the whole OS on an SD card then it should be safe for you to use it.

Yes, that's what I am sure, I may not needed system packages every time even I made updates with minor fixes.

```a.exclude_system_libraries(list_of_exceptions=['libexpat*', '*krb*'])```
I saw this yesterday and I could not understand better how to write on my below commandline.

Will it be like this ?

```
--exclude_system_libraries \
or
--exclude_system_libraries = list_of_exceptions=['numpy *', '*IPython *'] \
```
Is the **list_of_exceptions** mandatory ?

```
sudo pyinstaller \
    --name "${app_starter_name}" \
    --onefile \       
    --clean \
    --exclude-module _tkinter \
    --exclude-module PyQt4 \
    --exclude-module PyQt5 \
    --exclude-module numpy \
    --exclude-module IPython \
    --workpath build --distpath dist \
    main.py
```

**Reply by bwoodsend** [reply, marked as answer]
It's not a command line option. You have to switch to [using the spec file](https://pyinstaller.org/en/stable/spec-files.html#using-spec-files). And please never use sudo with PyInstaller.

**Reply by gaamaaresosa** [reply]
> It's not a command line option. 

Ok thanks again, I will try that too.

> please never use sudo with PyInstaller

Oh its a News for me. Thanks !
